Doc/c-api/monitoring.rst

```diff
@@ -0,0 +1,160 @@
+.. highlight:: c
+
+.. _monitoring:
+
+Monitorong C API
+================
+
+Added in version 3.13.
+
+An extension may need to interact with the event monitoring system. Subscribing
+to events and registering callbacks can be done via the Python API exposed in
+:mod:`sys.monitoring`.
+
+Generating Execution Events
+===========================
+
+The functions below make it possible for an extension to fire monitoring
+events as it emulates the execution of Python code. Each of these functions
+accepts a ``PyMonitoringState`` struct which contains concise information
+about the activation state of events, as well as the event arguments, which
+include a ``PyObject*`` representing the code object, the instruction offset
+and sometimes additional, event-specific arguments (see :mod:`sys.monitoring`
+for details about the signatures of the different event callbacks).
+The ``codelike`` argument should be an instance of :class:`types.CodeType`
+or of a type that emulates it.
+
+The VM disables tracing when firing an event, so there is no need for user
+code to do that.
+
+It is expected that monitoring functions are not called with an exception set,
+except for those which are firing exception-related events.
```

Comment:
`Py_THROW` is exception-related, but with this PR, calling `PyMonitoring_FirePyThrowEvent` with an exception set triggers an assertion error.

What's the expected behaviour -- *should* the user wrap the calls in `PyErr_GetRaisedException`/`PyErr_SetRaisedException`?

Reply:
I think we should probably say that the exception should not be set even for exception-related events (the exception should be an arg). @markshannon do you agree?

Reply:
That means users typically need to call `PyErr_GetRaisedException` & `PyErr_SetRaisedException` themselves, outside the internal `_PYMONITORING_IF_ACTIVE` check.

Reply:
We could do it for them.